### ISSUE TITLE
eBPF tracking causes panic on Ubuntu kernel 4.4.0-119

### ISSUE THREAD

**Issue by leth** [weaveworks/scope#3131]
Full kernel version: `4.4.0-119-generic #143-Ubuntu SMP`
We were previously running `4.4.0-116-generic #140-Ubuntu SMP` with no problems

## Workaround:
Disable eBPF connection tracking with `--probe.ebpf.connections=false`

## Panic details
[panic.txt](https://github.com/weaveworks/scope/files/1879867/panic.txt)

**Comment by dlespiau** [comment]
It'd be interesting to also have to known to work version so one can look at the list of commits that found their way in.

**Comment by rade** [comment]
@alban @iaguis any clues what might be going wrong here?

**Comment by alban** [comment]
Summary of the stack:
```
- bpf_map_lookup_elem
- bpf_prog_run
- fd_install
- do_sys_open
```

So the BPF program tries to read a eBPF map because of a open() syscall in the `scope` process:
https://github.com/weaveworks/tcptracer-bpf/blob/fecaba5/tcptracer-bpf.c#L873
```
	exists = bpf_map_lookup_elem(&fdinstall_pids, &tgid);
```

The message "BUG: unable to handle kernel paging request at 0000000063222428" might be an invalid pointer in the `bpf_map_lookup_elem()` function:
https://github.com/torvalds/linux/blob/v4.4/kernel/bpf/helpers.c#L41
```
	value = map->ops->map_lookup_elem(map, key);
```

Maybe `map` or `map->ops` is an invalid pointer? No idea why that would happen right now... is it 100% reproducible?

Given the timestamps, it seems to happen very soon after the machine boots, so probably during the initialization of Scope.

**Comment by rade** [comment]
> Given the timestamps, it seems to happen very soon after the machine boots, so probably during the initialization of Scope.

Yes. We've had a couple of machines in a k8s cluster in reboot loops since updating to the `4.4.0-119-generic #143-Ubuntu SMP` kernel.

**Comment by alban** [comment]
I see one [patch](https://launchpadlibrarian.net/363331633/linux-lts-xenial_4.4.0-116.140~14.04.1_4.4.0-119.143~14.04.1.diff.gz) on bpf maps in that new kernel (found from [here](https://www.ubuntuupdates.org/package/core/trusty/main/updates/linux-image-4.4.0-119-generic)):
https://pastebin.ubuntu.com/p/y58KDXwTMn/

Maybe it's a kernel patch that was badly backported to the Ubuntu 4.4 kernel?

**Comment by leth** [comment]
To apply the `--probe.ebpf.connections=false` argument to weave scope, you will need to edit the kubernetes daemonset manifest for weave scope. You can find it either in the `weave`, or `kube-system` namespace.

**Comment by JuneZhao** [comment]
@leth  thank you, one thing I am not clear is that where does this weave scope come from? I didnt create it by myself and I dont k8s ever created it previously? Does this weave scope comes with this kernel only? And it only applies to k8s?

**Comment by leth** [comment]
Weave scope is not part of the kernel, it uses the `eBPF` kernel feature.
Weave scope can be either manually installed on your system, or installed as part of connecting your system to Weave Cloud.
Weave scope can run on natively on your system, or within docker, or within kubernetes.

We have only seen the crash I posted under kernel `4.4.0-119-generic #143-Ubuntu SMP` so far.

Your report says "scope Not tainted" I'm not a kernel debugging expert, but this might mean that weave scope is not at fault?

**Comment by bboreham** [comment]
I tried it with just Scope, no Kubernetes or other stuff:
Before:
```
$ uname -a
Linux bryan-dev2 4.4.0-101-generic #124-Ubuntu SMP Fri Nov 10 18:29:59 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
$ scope launch
ebba7e65f5940c0eca410ca8f3d89f4839963ea5d1f0d52f79f5325a8762b50a
Scope probe started
Weave Scope is listening at the following URL(s):
  * http://10.240.0.2:4040/
```

After:
```
$ uname -a
Linux bryan-dev2 4.4.0-119-generic #143-Ubuntu SMP Mon Apr 2 16:08:24 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
$ scope launch
0a3b2a7a3580f2100f2374e334eea104a1f6f3371659de905db99c98942b1257
Scope probe started
ERRO[0002] error getting events from daemon: unexpected EOF 
sh: ../sysdeps/nptl/fork.c:156: __libc_fork: Assertion `THREAD_GETMEM (self, tid) != ppid' failed.
Aborted
[machine is unresponsive]
```

**Comment by schu** [comment]
Pull request for a workaround (fall back to procfs on affected kernels): https://github.com/weaveworks/scope/pull/3141

Ubuntu bug report: https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1763454

**Comment by putz612** [comment]
If you are getting bit by this you can block this specific kernel version using apt preferences until Ubuntu releases a new version.  That way you can apt-get update with without worrying about accidentally installing it.

> #cat /etc/apt/preferences

```
Package: linux*
Pin: version 4.4.0-119*
Pin-Priority: -1
```

**Comment by schu** [comment]
Update: Ubuntu has a fix in `Ubuntu 4.4.0-123.147-generic 4.4.128` ([`xenial-proposed`](https://wiki.ubuntu.com/Testing/EnableProposed)) that should be included in the next update cycle (as far as I can tell).

**Comment by michaelajr** [comment]
I wonder if this issue is causing my problems. I went from a Debian K8s cluster in AWS to an Ubuntu K8s cluster in AWS. Fresh cluster. Newest Ubuntu Image. Applied the WeaveScope Helm Chart - and BAM! All my hosts went out of service in the ELB (Instance status checks in AWS were failing too). When they came back in service - and I was able to do a `kubectl get pods` I saw all the WeaveScope pods in a crash loop backoff.  Only change is Ubuntu. I know this is comment has no log or any real info. I tore down the cluster, but will post back with more info while I debug this today.
